### PR TITLE
Infinite scroll for recent books page

### LINKED ISSUE #355: Implement infinite scroll / auto-loading for list views
Instead of requiring the user to click the "load more" link, let's automatically load more when scroll position reaches the bottom.

Some inspiration: https://elixircasts.io/infinite-scroll-with-liveview

List views that should be affected:

* [ ] Library recent books list
* [ ] Series books list
* [ ] Search results list

Lists that need some consideration:

These lists are embedded in pages with multiple lists and don't really have a good "scroll bottom".

* [ ] Continue Listening
* [ ] Authored / Narrated books

### PR DESCRIPTION
**Pull request by lubien**
I've created this draft PR so we can discuss how to continue with this.

Close #355

### CODE REVIEW

File: lib/ambry_web/components.ex [hunk 1]
```diff
@@ -474,8 +474,10 @@ defmodule AmbryWeb.Components do
       assigns
       |> assign_new(:show_load_more, fn -> false end)
       |> assign_new(:load_more, fn -> {false, false} end)
+      |> assign_new(:infinite_scroll, fn -> {false, false} end)
 
     {load_more, target} = assigns.load_more
+    {infinite_scroll_load_more, infinite_scroll_target} = assigns.infinite_scroll
```
Comment (lubien) [inline lib/ambry_web/components.ex:480]: For now I've added this so we don't mess with other usages but we might remove the previous load more if needed

**Comment by lubien** [comment]
Im not sure I know where to find those:

> Series books list
> Search results list

I do see the search one but doesn't seem to paginate so I'm not sure Im on the right place.

Can you help me pointing out where should I look at? cc @doughsay

**Comment by doughsay** [comment]
Hey @lubien!  Thanks so much for taking an interest and for your contribution.  I can't take too close a look at this right now, but I'll do a more thorough review later.

The series books list is in this live view: https://github.com/ambry-app/ambry/blob/main/lib/ambry_web/live/series_live/show.ex
But as I'm looking at it I don't think it's paginated at all, it loads all series books up front.

The search results page, you're right isn't paginated, it hard-codes the first 10 of each type and just aggregates them together. There's another ticket to switch the search results page to the newer version 2 of search here: https://github.com/ambry-app/ambry/issues/356

So maybe the best thing for now is to not worry about these other two places.

**Comment by lubien** [comment]
> There's another ticket to switch the search results page to the newer version 2 of search here: #356

I'm looking into this one right now, I might assign myself too!

> So maybe the best thing for now is to not worry about these other two places.

Agreed. I'll wait on your feedback before going forward, no rush at all

**Comment by doughsay** [comment]
Just trying it out now, looks like it works the first time, but then not after that anymore.  If you're running the project locally, the factories are available in dev mode.  Try something like this in an iex shell: `Ambry.Factory.insert_list(100, :media)` and it will just insert a bunch of dummy books.  It loads the initial batch, and then when you scroll it calls `load_mode` once, but then after it never calls it again.  Haven't dug too deeply into why yet.

File: lib/ambry_web/components.ex [hunk 2]
```diff
@@ -514,7 +516,11 @@ defmodule AmbryWeb.Components do
         </div>
       <% end %>
 
-      <%= if @show_load_more do %>
+      <%= if infinite_scroll_load_more do %>
+        <div id="infinite-scroll-marker" phx-hook="infiniteScroll" data-page={0} data-target={infinite_scroll_target}></div>
```
Comment (doughsay) [inline lib/ambry_web/components.ex:520]: I think the problem is the hard-coded `data-page={0}` here, looks like this needs to be dynamically updated based on what page was last loaded.

**Comment by lubien** [comment]
That's weird I slowly scrolled here and got two load more events.

I will take a look

File: lib/ambry_web/components.ex (hunk 2 above)
Comment (lubien) [reply lib/ambry_web/components.ex]: That was it :)

**Comment by lubien** [comment]
I've simplified things a bit and should work just fine now.

**Comment by doughsay** [comment]
@lubien great stuff, and thanks for contributing!  Works as advertised!

**Pull request merged**